### PR TITLE
[ISSUE-551] Support proactive removal for OFFLINE drive

### PR DESCRIPTION
**Pull request by safronovD**
### Issue #551 

- fix incorrect behavior on `health=bad` annotation for OFFLINE drive
- skip partition deletion if volume in MISSING
- locate Node instead of drive if it is OFFLINE

## PR checklist
- [x] Add link to the issue
- [x] Choose PR label
- [ ] New unit tests added
- [ ] Modified code has meaningful comments
- [ ] All TODOs are linked with the issues
- [ ] All comments are resolved

## Testing
```
# Pvc bound, volume created
[root@servicenode ~]# kubectl get volume
NAME                                       SIZE            STORAGE CLASS   HEALTH   CSI_STATUS   LOCATION                               NODE
pvc-25d3431e-03d5-48d1-9bf8-0cfa9a3c980d   8001000000000   HDD             GOOD     PUBLISHED    e1d1dfed-22bb-463e-95fa-7a32d49abd60   85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8
[root@servicenode ~]# kubectl get pvc
NAME         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
www-webi-0   Bound    pvc-25d3431e-03d5-48d1-9bf8-0cfa9a3c980d   7452Gi     RWO            csi-baremetal-sc   87s
[root@servicenode ~]# kubectl get drive -o wide | grep e1d1dfed-22bb-463e-95fa-7a32d49abd60
e1d1dfed-22bb-463e-95fa-7a32d49abd60   8001000000000   HDD    GOOD     ONLINE    IN_USE            /dev/sdh   VDH147TD           17     85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8

# Disk goes to OFFLINE
[root@servicenode ~]# kubectl get drive -o wide | grep e1d1dfed-22bb-463e-95fa-7a32d49abd60
e1d1dfed-22bb-463e-95fa-7a32d49abd60   8001000000000   HDD    UNKNOWN   OFFLINE   IN_USE            /dev/sdh   VDH147TD           17     85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8
[root@servicenode ~]# kubectl get volume
NAME                                       SIZE            STORAGE CLASS   HEALTH    CSI_STATUS   LOCATION                               NODE
pvc-25d3431e-03d5-48d1-9bf8-0cfa9a3c980d   8001000000000   HDD             UNKNOWN   PUBLISHED    e1d1dfed-22bb-463e-95fa-7a32d49abd60   85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8

# Annotate drive, wait DR start
[root@servicenode ~]# kubectl annotate drive e1d1dfed-22bb-463e-95fa-7a32d49abd60 health=bad
drive.csi-baremetal.dell.com/e1d1dfed-22bb-463e-95fa-7a32d49abd60 annotated
[root@servicenode ~]# kubectl get drive -o wide | grep e1d1dfed-22bb-463e-95fa-7a32d49abd60
e1d1dfed-22bb-463e-95fa-7a32d49abd60   8001000000000   HDD    BAD      OFFLINE   RELEASING            /dev/sdh   VDH147TD           17     85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8

# Perform disk removal
[root@servicenode ~]# kubectl annotate volume pvc-25d3431e-03d5-48d1-9bf8-0cfa9a3c980d release=done
volume.csi-baremetal.dell.com/pvc-25d3431e-03d5-48d1-9bf8-0cfa9a3c980d annotated
[root@servicenode ~]# kubectl get volume -o wide
NAME                                       SIZE            STORAGE CLASS   HEALTH   CSI_STATUS   OP_STATUS   USAGE      TYPE   LOCATION                               NODE
pvc-25d3431e-03d5-48d1-9bf8-0cfa9a3c980d   8001000000000   HDD             BAD      PUBLISHED    MISSING     RELEASED   xfs    e1d1dfed-22bb-463e-95fa-7a32d49abd60   85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8
[root@servicenode ~]# kubectl get drive -o wide | grep e1d1dfed-22bb-463e-95fa-7a32d49abd60
e1d1dfed-22bb-463e-95fa-7a32d49abd60   8001000000000   HDD    BAD      OFFLINE   RELEASED            /dev/sdh   VDH147TD           17     85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8
[root@servicenode ~]# kubectl annotate drive e1d1dfed-22bb-463e-95fa-7a32d49abd60 removal=ready
drive.csi-baremetal.dell.com/e1d1dfed-22bb-463e-95fa-7a32d49abd60 annotated
[root@servicenode ~]# kubectl get drive -o wide | grep e1d1dfed-22bb-463e-95fa-7a32d49abd60
e1d1dfed-22bb-463e-95fa-7a32d49abd60   8001000000000   HDD    BAD      OFFLINE   REMOVING            /dev/sdh   VDH147TD           17     85cf358f-3b5b-4c76-961d-9fb6f0e2ebe8
[root@servicenode ~]# kubectl delete pvc www-webi-0 --wait=false
persistentvolumeclaim "www-webi-0" deleted
[root@servicenode ~]# kubectl delete po webi-0
pod "webi-0" deleted

# Disk removed
[root@servicenode ~]# kubectl get drive -o wide e1d1dfed-22bb-463e-95fa-7a32d49abd60
Error from server (NotFound): drives.csi-baremetal.dell.com "e1d1dfed-22bb-463e-95fa-7a32d49abd60" not found

# Pod running, pvc bound
[root@servicenode ~]# kubectl get pvc
NAME         STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS       AGE
www-webi-0   Bound    pvc-fd4b9b9f-b1d5-46c3-a1e0-4267f634eb0e   7452Gi     RWO            csi-baremetal-sc   12m
[root@servicenode ~]# kubectl get po webi-0
NAME     READY   STATUS    RESTARTS   AGE
webi-0   1/1     Running   0          13m
```

### CODE REVIEW

File: pkg/node/volumemgr.go [hunk 1]
```diff
@@ -474,33 +474,47 @@ func (m *VolumeManager) handleRemovingStatus(ctx context.Context, volume *volume
...
 	})
 
-	var (
-		err       error
-		newStatus string
-	)
-	if err = m.getProvisionerForVolume(&volume.Spec).ReleaseVolume(volume.Spec); err != nil {
+	newStatus, err := m.performVolumeRemoving(ctx, volume)
+	if err != nil && newStatus == "" {
+		return ctrl.Result{Requeue: true}, err
```
Comment (mishoyama) [inline pkg/node/volumemgr.go:479]: Is it possible to address `codecov` warnings?

**Review by mishoyama — APPROVED** [review]

**Review by rkiselenko — APPROVED** [review]

File: pkg/crcontrollers/drive/drivecontroller.go [hunk 2]
```diff
@@ -321,6 +319,28 @@ func (c *Controller) checkAndPlaceStatusRemoved(drive *drivecrd.Drive) bool {
 	return false
 }
 
+func (c *Controller) handleDriveUsageRemoving(ctx context.Context, log *logrus.Entry, drive *drivecrd.Drive) (uint8, error) {
+	volumes, err := c.crHelper.GetVolumesByLocation(ctx, drive.Spec.UUID)
+	if err != nil {
+		return ignore, err
+	}
+	if c.checkAllVolsRemoved(volumes) {
```
[the remark this answers is not in the document]
Comment (safronovD) [reply pkg/crcontrollers/drive/drivecontroller.go]: Got it, thanks

File: pkg/node/volumemgr.go (hunk 1 above)
Comment (safronovD) [reply pkg/node/volumemgr.go:479]: add UT for overridden health

Comment (safronovD) [reply pkg/node/volumemgr.go:479]: But it's need to fail k8sclient.Update function in the test above. Unfortunately, I can't find any simple way to do it, only creating another interface. Did you ever met references for this?

**Review by yurkov-anton — APPROVED** [review]

**Pull request merged**